lib/class-wp-rest-block-types-controller.php

```diff
@@ -0,0 +1,326 @@
+		}
+
+		$data = $this->prepare_item_for_response( $block_type, $request );
+
+		return rest_ensure_response( $data );
+	}
+
+	/**
+	 * Prepares a block type object for serialization.
+	 *
+	 * @param stdClass        $block_type block type data.
+	 * @param WP_REST_Request $request    Full details about the request.
+	 *
+	 * @return WP_REST_Response block type data.
+	 */
+	public function prepare_item_for_response( $block_type, $request ) {
+
+		$fields = $this->get_fields_for_response( $request );
+		$data   = array();
+		if ( in_array( 'name', $fields, true ) ) {
+			$data['name'] = $block_type->name;
+		}
+
+		if ( in_array( 'attributes', $fields, true ) ) {
+			$data['attributes'] = $block_type->get_attributes();
+		}
+
+		if ( in_array( 'is_dynamic', $fields, true ) ) {
+			$data['is_dynamic'] = $block_type->is_dynamic();
+		}
+
+		$extra_fields = array( 'editor_script', 'script', 'editor_style', 'style' );
+		foreach ( $extra_fields as $extra_field ) {
+			if ( in_array( $extra_field, $fields, true ) ) {
+				$data[ $extra_field ] = (string) $block_type->$extra_field;
+			}
+		}
+
+		$context = ! empty( $request['context'] ) ? $request['context'] : 'view';
+		$data    = $this->add_additional_fields_to_object( $data, $request );
+		$data    = $this->filter_response_by_context( $data, $context );
+
+		$response = rest_ensure_response( $data );
+
+		$response->add_links(
+			array(
+				'collection'              => array(
+					'href' => rest_url( sprintf( '%s/%s', $this->namespace, $this->rest_base ) ),
+				),
+				'https://api.w.org/items' => array(
+					'href' => rest_url( sprintf( '%s/%s/%s', $this->namespace, $this->rest_base, $block_type->name ) ),
+				),
+			)
+		);
+
+		/**
+		 * Filters a block type returned from the REST API.
+		 *
+		 * Allows modification of the block type data right before it is returned.
+		 *
+		 * @param WP_REST_Response $response   The response object.
+		 * @param object           $block_type The original block type object.
+		 * @param WP_REST_Request  $request    Request used to generate the response.
+		 */
+		return apply_filters( 'rest_prepare_block_type', $response, $block_type, $request );
+	}
+
+	/**
+	 * Retrieves the block type' schema, conforming to JSON Schema.
+	 *
+	 * @return array Item schema data.
+	 */
+	public function get_item_schema() {
+		if ( $this->schema ) {
+			return $this->add_additional_fields_schema( $this->schema );
+		}
+
+		$schema = array(
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
+			'title'      => 'block_type',
+			'type'       => 'object',
+			'properties' => array(
+				'name'          => array(
+					'description' => __( 'The title for the block type.', 'gutenberg' ),
+					'type'        => 'string',
+					'context'     => array( 'embed', 'view', 'edit' ),
+					'readonly'    => true,
+				),
+				'attributes'    => array(
+					'description' => __( 'Block attributes.', 'gutenberg' ),
+					'type'        => 'object',
```

Comment:
This field is a big part of the problem I have with the lack of server registry for blocks. 

This is probably _the_ most important part of blocks. Blocks should be able to register their attributes on the server, what _kind_ of attribute it is, what values should be allowed to be saved (sanitization/validation rules), etc. 

Headless applications can't be reliably built if the payload is unpredictable, which is my main point with #2751 in the first place. If the server is telling the client: "this block may or may not have some object with arbitrary, unpredictable shape as attributes", what use is it really to the client?

Reply:
@spacedmonkey Is this something to be concerned with for the implementation here? 

I may be reading it wrong, but: The schema for the endpoint is generalized of "some object", but that's because the actual response _is_ the attribute schema.

So, it's fine? Or am I missing something?

Reply:
I don't see the issue here. Attributes are defined per block so defined here.

Reply:
As a consumer of an API, it's important to know what data you are going to get.

If the consumer just knows "this field can literally be anything" it's very difficult to build an application. You have to build your application to prepare for "anything".

It would be good for the REST Schema to be able to let the consumers know _before_ execution what they can expect here. 

It would be like defining the `/post` endpoint to return an object. That's not helpful. Instead, we define in the Schema that it has an ID and a Title and a Date, etc. 

This allows consumers to know what to expect before execution. 

Telling a consumer "this will be an object" isn't really helpful unless you can also tell the consumer what the shape of the object will be, or the possible variations in shape. 

This is the reason the WP REST API has a Schema API in the first place. To allow client applications to be able to predict the payloads before executing them. Returning objects with no definition takes that away.

Reply:
I think this is a misunderstanding of the endpoint here. This schema is describing that the API will return attributes which are a JSON schema. We've provided the most accurate schema definition that we can which is that each attribute is an `object` since it is itself a JSON schema definition.

The endpoint returns that rich schema you are talking about.

```json
{
  "attributes": {
    "align": {
      "type": "string",
      "enum": [
        "left",
        "center",
        "right",
        "wide",
        "full"
      ]
    },
    "className": {
      "type": "string"
    },
    "displayAsDropdown": {
      "type": "boolean",
      "default": false
    },
    "showPostCounts": {
      "type": "boolean",
      "default": false
    },
    "layout": {
      "type": "string"
    }
  }
}
```

Reply:
Each blocks attribututes are different as attributes is use a store of data used by the block. There is no way to describe this data as is on a block by block basis.

Reply:
The Schema should expose what's _possible_, and the execution should show what _is_.

What's possible is for this block to have an attribute of align with the possible values of left/center/right/wide/full, a className that's a string, etc. 

That should be exposed in the Schema. 

The values of those attributes `align: right` `className: 'something'` should be what's returned from a request. 

We know what attributes are possible on blocks. 

We don't have to fetch `/post/fields` to know that a Post has an id field that's an integer, a title field that's a string, etc. That is all defined in the Schema and we, as decoupled application developers, can use introspection to know what is possible to be returned by the API. 

The attributes should be exposed as part of the Schema, as they are expressing what's possible.

Reply:
@jasonbahl you seem to be under the impression that this is a blocks controller, it isn't, it is a block type controller. Similar to the post types controller. The values of those attributes is not `align: right className: 'something'` and that is not what is returned by this request.

This returns the definition for the block type. If you look at the response you'll see this returns the attributes the block type declares, not the blocks themselves.

```json
{
  "attributes": {
    "align": {
      "type": "string",
      "enum": [
        "left",
        "center",
        "right",
        "wide",
        "full"
      ]
    },
    "className": {
      "type": "string"
    },
    "displayAsDropdown": {
      "type": "boolean",
      "default": false
    },
    "showPostCounts": {
      "type": "boolean",
      "default": false
    },
    "layout": {
      "type": "string"
    }
  },
  "is_dynamic": true,
  "name": "core\/archives",
  "category": "widgets",
  "editor_script": "",
  "script": "",
  "editor_style": "",
  "style": "",
  "supports": {
    "align": true,
    "html": false
  },
  "title": "",
  "icon": "",
  "description": "",
  "keywords": [],
  "parent": [],
  "styles": [],
  "text_domain": "",
  "_links": {
    "collection": [
      {
        "href": "http:\/\/localhost:8889\/index.php?rest_route=\/__experimental\/block-types"
      }
    ],
    "self": [
      {
        "href": "http:\/\/localhost:8889\/index.php?rest_route=\/__experimental\/block-types\/core\/archives"
      }
    ],
    "up": [
      {
        "href": "http:\/\/localhost:8889\/index.php?rest_route=\/__experimental\/block-types\/core"
      }
    ],
    "wp:render-block": [
      {
        "href": "http:\/\/localhost:8889\/index.php?rest_route=%2Fwp%2Fv2%2Fblock-renderer%2Fcore%2Farchives&context=edit"
      }
    ],
    "curies": [
      {
        "name": "wp",
        "href": "https:\/\/api.w.org\/{rel}",
        "templated": true
      }
    ]
  }
}
```

This would be like expecting the post types controller to have a schema that lists all the possible meta fields.

Reply:
Also worth noting, this is more like the register_meta api in WP Core. This registers a key name, it's type and some callbacks. Just becuase you have an array of values in meta, there is no guideline on how that data is created. Say for example of array of possible values should be implemented with a select menu or a radio button in edit for example. That is what a fields api is for now. To map the raw data to an implemention. Gutenberg does not have this and this is not the PR that this would implement this. This PR is just simply to explose existing data regsitered in the PHP so that gutenberg and other clients can get this data.

Reply:
> Each blocks attribututes are different as attributes is use a store of data used by the block. There is no way to describe this data as is on a block by block basis.

That was the main point of my comments in #2751 and having Gutenberg blocks be registered on the server in the first place. So we wouldn't have to say "There is no way to describe this data". . .the data would be described in the block registry, so we could expose that to a Schema.

Reply:
WPGraphQL for Gutenberg and Gutenberg Fields Middleware are good examples of this concept, where attributes can be exposed by an API. 

Sad to see #2751 closed without actually addressing this issue.

Reply:
@TimothyBJacobs reached out to me on slack to clarify some things. While some things are not ideal (in my mind) this is for sure a step in a much better direction. I look forward to how this continues to improve to allow Gutenblocks to be useful to other consumers. 

Thanks for your work on this @spacedmonkey!